File: src/audioWfc/constraints/AscendingMelodySoftConstraint.java
package audioWfc.constraints;

import audioWfc.wfc.HigherValues;
import audioWfc.musicTheory.OctavedNote;
import audioWfc.wfc.Tile;
import audioWfc.constraints.concepts.SoftConstraint;

public class AscendingMelodySoftConstraint extends SoftConstraint<OctavedNote> {
    public AscendingMelodySoftConstraint(double factor){
        super(factor);
    }

    @Override
    public double weight(Tile<OctavedNote> item, HigherValues higherValues) {
        OctavedNote note = item.getValue();

        Tile<OctavedNote> prev = item.getPrev();
        Tile<OctavedNote> next = item.getNext();

        double out = 0d;

        if(prev.isCollapsed() && checkAscending(prev.getValue(), note)){
            out += factor;
        }

        if(next.isCollapsed() && checkAscending(note, next.getValue())){
            out += factor;
        }

        return out;
    }

    private boolean checkAscending(OctavedNote first, OctavedNote second){
        return second.MIDIValue() > first.MIDIValue();
    }
}

File: src/wfc/constraints/AscendingMelodySoftConstraint.ts
import { OctavedNote } from "../../music_theory/Note"
import { Tile } from "../Tile"
import { SoftConstraint } from "./concepts/Constraint"
import { noteConstraintTypeToName } from "./constraintUtils"

export const AscendingMelodySoftConstraintInit = {
	type: "AscendingMelodySoftConstraint" as const,
	bonus: 1,
	validByDefault: true as const,
	reachOverPrev: true,
	reachOverNext: true
}

export type AscendingMelodySoftConstraintIR =
	typeof AscendingMelodySoftConstraintInit
export class AscendingMelodySoftConstraint extends SoftConstraint<OctavedNote> {
	name = noteConstraintTypeToName.get(
		AscendingMelodySoftConstraintInit.type,
	)!.name as string



	constructor(bonus: number, private reachOverPrev: boolean, private reachOverNext: boolean) {
		super(bonus)
	}

	weight(tile: Tile<OctavedNote>): number {
		const note = tile.getValue()
		const prev = tile.getPrev(this.reachOverPrev)
		const next = tile.getNext(this.reachOverNext)

		let out = 0

		if (prev.isCollapsed() && this.isAscending(prev.getValue(), note))
			out += this.bonus
		if (next.isCollapsed() && this.isAscending(note, next.getValue()))
			out += this.bonus

		return out
	}

	private isAscending(first: OctavedNote, second: OctavedNote): boolean {
		return first.toMIDIValue() < second.toMIDIValue()
	}
}
